crates/core/src/client/mod.rs

```diff
@@ -499,12 +488,68 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
     }
 
     /// Returns the estimated gas for a transaction
-    async fn estimate_gas(
-        &self,
-        _call_request: CallRequest,
-        _block_number: Option<BlockId>,
-    ) -> Result<U256, EthApiError<P::Error>> {
-        todo!();
+    async fn estimate_gas(&self, request: CallRequest, block_id: BlockId) -> Result<U256, EthApiError<P::Error>> {
+        let chain_id: u64 = request
+            .chain_id
+            .ok_or_else(|| EthApiError::MissingParameterError("chain_id for estimate_gas".into()))?
+            .low_u64();
+
+        let from = request.from.ok_or_else(|| EthApiError::MissingParameterError("from for estimate_gas".into()))?;
+        let nonce: Felt252Wrapper = self.nonce(from, block_id).await?.try_into()?;
+
+        let gas_limit: Felt252Wrapper = request.gas.unwrap_or_else(|| U256::from(1_000_000)).try_into()?;
+        let max_fee_per_gas: Felt252Wrapper =
+            request.max_fee_per_gas.unwrap_or_else(|| U256::from(BASE_FEE_PER_GAS)).try_into()?;
+        let max_priority_fee_per_gas: Felt252Wrapper =
+            request.max_priority_fee_per_gas.unwrap_or_else(|| U256::from(MAX_PRIORITY_FEE_PER_GAS)).try_into()?;
+
+        let to = request.to.map_or(TransactionKind::Create, TransactionKind::Call);
+
+        let value: Felt252Wrapper = request.value.unwrap_or_else(|| U256::from(0)).try_into()?;
+
+        let data = request.data.unwrap_or_default();
+
+        let tx = Transaction::Eip1559(TxEip1559 {
+            chain_id,
+            nonce: nonce.clone().try_into()?,
+            gas_limit: gas_limit.try_into()?,
+            max_fee_per_gas: max_fee_per_gas.try_into()?,
+            max_priority_fee_per_gas: max_priority_fee_per_gas.try_into()?,
+            to,
+            value: value.try_into()?,
+            access_list: AccessList(vec![]),
+            input: data,
+        });
+
+        let starknet_block_id: StarknetBlockId = EthBlockId::new(block_id).try_into()?;
+        let block_number = self.get_block_number(&starknet_block_id).await?;
+
+        let sender_address = self.compute_starknet_address(from, &starknet_block_id).await?;
+
+        let mut data = vec![];
+        tx.encode_with_signature(&Signature::default(), &mut data, false);
+        let mut eth_calldata = data.into_iter().map(FieldElement::from).collect::<Vec<_>>();
+
+        let mut calldata = vec![
```

Comment:
isn't there any utils to create such a `Call`? I thought we added some recently

Reply:
We don't have a wrapper for `Call`, we have a wrapper for `Calls=Vec<Call>`. I can add a wrapper for `Call`, which would then convert a `Call` into a `Vec<FieldElement>`, what do you think?

Reply:
and can't we use `let calldata = raw_starknet_calldata(self.kakarot_address(), bytes);` as in `send_transaction`?

Reply:
Changed to use `raw_starknet_calldata`. Also updated `raw_starknet_calldata` because the implementation was very weird.